### PR TITLE
[3875][ADD] base_api_connection

### PR DESCRIPTION
**Pull request by AungKoKoLin1997**
[3875](https://www.quartile.co/web#id=3875&menu_id=505&cids=3&action=1457&model=project.task&view_type=form)

### CODE REVIEW

**Review by yostashiro — CHANGES_REQUESTED** [review]
As talked earlier, please create another bridge module `api_connection_key_encryption` (AGPL-3) to take care of the encryption/decription parts.

File: base_api_connection/models/api_call_mixin.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        endpoint=None,
+        custom_headers=None,
+        params=None,
+        json=None,
+        http_method="get",
+    ):
+        config = self.env["api.config"].search(
+            [("external_system", "=", external_system), ("code", "=", code)], limit=1
+        )
+        if not config:
```
Comment (yostashiro) [inline base_api_connection/models/api_call_mixin.py:34]:
```suggestion
        if not config:
            _logger.warning("No API config record found.")
```

**Review by yostashiro — COMMENTED** [review]
Partial review.

File: base_api_connection/README.rst [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+This module depends on the 'data_encryption' OCA
+module(https://github.com/OCA/server-env/blob/16.0/data_encryption).
+Therefore, please adhere to the configuration steps outlined in this
+module.
+
+Bug Tracker
+===========
+
+Bugs are tracked on `GitHub Issues <https://github.com/qrtl/axls-private/issues>`_.
```
Comment (kanda999) [inline base_api_connection/README.rst]:
```suggestion
Bugs are tracked on `GitHub Issues <https://github.com/qrtl/axls-oca/issues>`_.
```
Can you please create the README file again?

**Review by kanda999 — APPROVED** [review]
Code and Functional review

**Comment by AungKoKoLin1997** [comment]
I change the readme file type from `.md` to `.rst` because generating readme manually doesn't recognize `.md`.

File: base_api_connection/models/api_config.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+from odoo import _, api, fields, models
+from odoo.exceptions import ValidationError
+
+
+class ApiConfig(models.Model):
+    _name = "api.config"
+    _description = "API Configuration"
+
+    name = fields.Char(required=True)
+    code = fields.Char(required=True)
```
Comment (yostashiro) [inline base_api_connection/models/api_config.py]:
```suggestion
    code = fields.Char(required=True, help="Expected to act as an identifier of the API configuration record along with External System.")
```

File: base_api_connection/readme/DESCRIPTION.rst [hunk 4]
```diff
@@ -0,0 +1 @@
+This module facilitates API connections between Odoo and other system. It is not usable on its own; rather, it is a low-level module intended to serve as a base for others. An example of such a module is 'project_task_capture'.
```
Comment (yostashiro) [inline base_api_connection/readme/DESCRIPTION.rst]:
```suggestion
This module facilitates API connections between Odoo and other web services. It is not usable on its own; rather, it is a low-level module intended to serve as a base for others. An example of such a module is 'project_task_capture'.
```

File: base_api_connection/models/api_config.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    def _check_code(self):
+        for record in self:
+            duplicate_rec = self.search(
+                [
+                    ("code", "=", record.code),
+                    ("id", "!=", record.id),
+                ],
+                limit=1,
+            )
+            if duplicate_rec:
```
Comment (yostashiro) [inline base_api_connection/models/api_config.py]:
```suggestion
            existing_rec = self.search(
                [("code", "=", record.code), ("id", "!=", record.id)],
            )
            if existing_rec:
```

**Review by yostashiro — APPROVED** [review]
Just minor comments.

**Pull request merged**